text-spacing: text-autospace: handle element boundary spacing on IFC (layout part)

vitorroriz opened the pull request:
#### 857fd6ae4fe6ff242b4375eb3538dc2995e93434
<pre>
text-spacing: text-autospace: handle element boundary spacing on IFC (layout part)
<a href="https://bugs.webkit.org/show_bug.cgi?id=278531">https://bugs.webkit.org/show_bug.cgi?id=278531</a>
<a href="https://rdar.apple.com/133319434">rdar://133319434</a>

Reviewed by Alan Baradlay.

This patch is a step on handling the addition of extra spacing between relevant
characters as required by the text-autospace property [1].
The work will be divided into 2 parts: lauout and display. This patch is
responsible for display.

Up to know we are able to handle spacing between characters within an element.
We also want to handle the case of two adjacent characters that belong to different
elements.

For characters within an element we add spacing by manipulating the
correspondent&apos;s character&apos;s glyphs and advances. We always add spacing as
leading spacing. So if character &quot;A&quot; and &quot;B&quot; need spacing in between them,
this is added as a leading spacing of &quot;B&quot; by expanding its advance.width and
moving its origin.x by the required spacing size.

However, for element boundaries, specification [1] says:

&quot;At element boundaries, the amount of extra spacing introduced between
characters is determined by and rendered within the innermost element that
contains the boundary.&quot;

This means that for such a case, spacing has to be added between the elements.

We compute the necessary spacing to be &quot;placed&quot; in between element boundaries at
InlineItemsBuilder::build with computeInlineBoxBoundaryTextSpacingsIfNeeded.
At InlineItemsBuilder::computeInlineTextItemWidths we use such spacing to
contribute to the related InlineTextItem&apos;s width.

Important thing to notice:
Given the example:
&lt;span&gt;agua&lt;/span&gt;&lt;span&gt;水&lt;span&gt;
This would be represented by:
 InlineStartBox | InlineTextItem | InlineEndBox | InlineStartBox | InlineTextItem | InlineEndBox
 Being &quot;agua&quot; the content of the first InlineTextItem and &quot;水&quot; the content of the second InlineTextItem.

 Visually, the spacing necessary between &quot;agua&quot; and &quot;水&quot; has to appear between the first InlineEndBox
 and the second InlineStartBox. This means that such spacing should not visually contribute for
 the width of InlineTextItem. It might seem that we are doing this at this patch, however,
 this part is just addressing layout, and this information is sufficient for InlineContentBreaker to
 take such spacing in consideration. Later, on a future patch which will address &quot;display&quot;, we will
 then need to place this spacing in its correct visual order: it should me moved in front of the InlineStartBox
 and its value should be removed from the InlineTextItem&apos;s width, such that it doesn&apos;t get accounted for twice.

[1] <a href="https://www.w3.org/TR/css-text-4/#text-autospace-property">https://www.w3.org/TR/css-text-4/#text-autospace-property</a>

* Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
(WebCore::Layout::InlineContentCache::inlineBoxBoundaryTextSpacings const):
(WebCore::Layout::InlineContentCache::setInlineBoxBoundaryTextSpacings):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.h:
(WebCore::Layout::InlineFormattingContext::inlineContentCache const):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::build):
(WebCore::Layout::InlineItemsBuilder::adjustInlineBoxBoundaryTextSpacingsIfNeeded):
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.h:
* Source/WebCore/platform/text/TextSpacing.cpp:
(WebCore::TextAutospace::shouldApplySpacing const):
* Source/WebCore/platform/text/TextSpacing.h:

Canonical link: <a href="https://commits.webkit.org/282917@main">https://commits.webkit.org/282917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe3e2fc7ca39a82785fbbabd502eb8da1ed19983

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68659 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15242 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15521 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10526 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67702 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55932 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32618 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37381 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13307 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14115 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59309 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70364 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8581 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13141 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59323 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56017 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14265 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7098 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/765 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39813 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40890 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42073 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItem.h:
```diff
@@ -69,15 +71,18 @@ class InlineItem {
 protected:
     InlineLayoutUnit m_width { };
     unsigned m_length { 0 };
+    float m_offset { 0 };
```
please find a less generic name. I don't want people misuse this highly specific member.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItem.h:
```diff
@@ -109,6 +114,11 @@ inline void InlineItem::setWidth(InlineLayoutUnit width)
+{
+    m_offset = offset;
+}
+
```
I assume this is not an offset on _any_ type of inline item. please assert that we put this on the correct type of inline items.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
```diff
@@ -297,6 +297,9 @@ void Line::appendInlineBoxStart(const InlineItem& inlineItem, const RenderStyle&
     // This is really just a placeholder to mark the start of the inline box <span>.
     ++m_nonSpanningInlineLevelBoxCount;
     auto logicalLeft = lastRunLogicalRight();
+    // If startbox has an offset, it should contribute to its logical left, except if it is the first box in the line.
```
what's a startbox (this is a start of an inline box, emphasis on inline box) ? also comment should say why we are trimming here.

vitorroriz replied:
I wanted to reference `Type::InlineBoxStart` name. I will replace it by `InlineBoxStart`.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItem.cpp:
```diff
@@ -35,6 +35,7 @@ namespace Layout {
 struct SameSizeAsInlineItem {
     void* layoutBox;
     float width;
+    float offset;
```
this is quite an expansion for something this uncommon.

vitorroriz replied:
I actually could use `width` the width of InlineItem (I guess it is not used when the item is a `Type::InlineBoxStart` ?) to save a float, but `width` does not hold the right meaning of what we want to do. Do you think it would be better, or should we just find a better name for `offset`?

alanbaradlay replied:
inline boxes can have decoration(s) so width is actually used on them. I can think of 2 alternatives here.
- use a side structure for this data
- keep it part of the InlineTextItem width and transfer it over to outside of the inline box while building the line (but this is probably too involved and error prone)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -678,22 +678,66 @@ static inline bool canCacheMeasuredWidthOnInlineTextItem(const InlineTextBox& in
+    {
+        if (!m_unhandledBoxStart)
+            return;
+        auto textAutoSpace = inlineTextBox.parent().style().textAutospace();
```
maybe hold on to the parent style? (used in 2 places here)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -678,22 +678,66 @@ static inline bool canCacheMeasuredWidthOnInlineTextItem(const InlineTextBox& in
     return !inlineTextBox.hasPositionDependentContentWidth();
 }
 
+class TextAutospaceAdjuster {
```
I think I commented on it later already but let me put a generic observation here: _box start_ is not a valid term. please use inline box start instead.

vitorroriz replied:
Right, I was referencing `Type::InlineBoxStart`. I'll replace it. Thanks.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -678,22 +678,66 @@ static inline bool canCacheMeasuredWidthOnInlineTextItem(const InlineTextBox& in
 void InlineItemsBuilder::computeInlineTextItemWidths(InlineItemList& inlineItemList)
 {
-    TextSpacing::SpacingState spacingState;
+    TextAutospaceAdjuster textAutospaceAdjuster;
```
is this going to be a perf concern for content with no space adjustment at all?

vitorroriz replied:
I haven't measured a perceived performance for content with `text-autospace: no`. I'm measuring again for the most recent iteration of this patch.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
```diff
@@ -52,6 +54,7 @@ class InlineContentCache {
         void set(InlineItemList&&, ContentAttributes);
         void replace(size_t insertionPosition, InlineItemList&&, ContentAttributes);
         void shrinkToFit() { m_inlineItemList.shrinkToFit(); }
+        void setElementBoundaryTextSpacings(ElementBoundaryTextSpacings&& spacings) { m_elementBoundaryTextSpacings = WTFMove(spacings); }
```
maybe use a more specific term like InlineBoxBoundary? (element boundary would indicate a more generic approach where spacing may cross generic inline level elements)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -114,6 +114,58 @@ void InlineItemsBuilder::build(InlineItemPosition startPosition)
     };
     adjustInlineContentCacheWithNewInlineItems();
 
+    auto adjustElementBoundaryTextSpacings = [&] {
```
this function is getting too long and this is certainly not core functionality, could you move it to a dedicated function, please?

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -114,6 +114,58 @@ void InlineItemsBuilder::build(InlineItemPosition startPosition)
+    auto adjustElementBoundaryTextSpacings = [&] {
+        char32_t lastCharacterFromPreviousRun = 0;
+        unsigned lastCharacterDepth = 0;
+        unsigned currentCharacterDepth = 0;
```
IFC has adopted size_t usage.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
```diff
@@ -297,6 +298,16 @@ void Line::appendInlineBoxStart(const InlineItem& inlineItem, const RenderStyle&
+        if (auto elementBoundarySpacing = elementBoundarySpacings.find(indexOnInlineItemList); elementBoundarySpacing !=  elementBoundarySpacings.end())
+            logicalLeft += elementBoundarySpacing->value;
+    }
+
```
since this will certainly not work with run reordering, it might be better to not introduce this level of complexity for the non-bidi case only and represent the gap as logical width instead. -which means Line does not need to know about this at all (as we discussed offline, Line is just a representation of widths in logical order with no real horizontal placement).

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.h:
```diff
@@ -79,6 +79,8 @@ class InlineFormattingContext {
     const BoxGeometry& geometryForBox(const Box&, std::optional<EscapeReason> = std::nullopt) const;
     BoxGeometry& geometryForBox(const Box&, std::optional<EscapeReason> = std::nullopt);
 
+    const InlineContentCache& inlineContentCache() const { return m_inlineContentCache; }
```
Builders tend to communicate with each other by passing in objects as opposed to adding getters to generic public interfaces.

alanbaradlay replied:
Could you remove this please?

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
```diff
@@ -61,9 +64,12 @@ class InlineContentCache {
     private:
         ContentAttributes m_contentAttributes;
         InlineItemList m_inlineItemList;
+        ElementBoundaryTextSpacings m_elementBoundaryTextSpacings;
```
This looks to me like a side structure to InlineItems, so should probably go _next_ to InlineItems.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -114,6 +114,58 @@ void InlineItemsBuilder::build(InlineItemPosition startPosition)
+            unsigned boundaryDepth = std::min(currentCharacterDepth, lastCharacterDepth);
+            unsigned boundaryIndex = inlineBoxStartIndexesOnInlineItemsList.size() - 1 - (currentCharacterDepth - boundaryDepth);
+            ASSERT(boundaryIndex >= 0);
+            const RenderStyle& boundaryStyle = inlineItemList[boundaryIndex].layoutBox().parent().style();
```
"parent style" would be more informative as boundary is normally between 2 adjacent entries which makes it hard to reason about what style we are using here (also applies to 'boundary' here in general, e.g. not sure if boundary depth is descriptive enough)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.h:
```diff
@@ -72,6 +72,7 @@ class InlineItemsBuilder {
     bool m_contentRequiresVisualReordering { false };
     bool m_isTextAndForcedLineBreakOnlyContent { true };
     size_t m_inlineBoxCount { 0 };
+    bool m_hasSeenTextAutospace { !root().style().textAutospace().isNoAutospace() };
```
should this always be false? at construction time we certainly have not seen any content with autospace (unless this member actually means something slightly different)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -134,6 +135,60 @@ void InlineItemsBuilder::build(InlineItemPosition startPosition)
+        }
+
+        unsigned boundaryDepth = std::min(currentCharacterDepth, lastCharacterDepth);
+        unsigned boundaryIndex = inlineBoxStartIndexesOnInlineItemsList.size() - 1 - (currentCharacterDepth - boundaryDepth);
```
size_t?

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -134,6 +135,60 @@ void InlineItemsBuilder::build(InlineItemPosition startPosition)
+
+        unsigned boundaryDepth = std::min(currentCharacterDepth, lastCharacterDepth);
+        unsigned boundaryIndex = inlineBoxStartIndexesOnInlineItemsList.size() - 1 - (currentCharacterDepth - boundaryDepth);
+        ASSERT(boundaryIndex >= 0);
```
This assert seems redundant. unsigned can't have negative value. if you wanna assert for arithmetic overflow, maybe assert on the  statement itself.

vitorroriz committed:
text-spacing: text-autospace: handle element boundary spacing on IFC (layout part)
https://bugs.webkit.org/show_bug.cgi?id=278531
rdar://133319434

Reviewed by Alan Baradlay.

This patch is a step on handling the addition of extra spacing between relevant
characters as required by the text-autospace property [1].
The work will be divided into 2 parts: lauout and display. This patch is
responsible for display.

Up to know we are able to handle spacing between characters within an element.
We also want to handle the case of two adjacent characters that belong to different
elements.

For characters within an element we add spacing by manipulating the
correspondent's character's glyphs and advances. We always add spacing as
leading spacing. So if character "A" and "B" need spacing in between them,
this is added as a leading spacing of "B" by expanding its advance.width and
moving its origin.x by the required spacing size.

However, for element boundaries, specification [1] says:

"At element boundaries, the amount of extra spacing introduced between
characters is determined by and rendered within the innermost element that
contains the boundary."

This means that for such a case, spacing has to be added between the elements.

We compute the necessary spacing to be "placed" in between element boundaries at
InlineItemsBuilder::build with computeInlineBoxBoundaryTextSpacingsIfNeeded.
At InlineItemsBuilder::computeInlineTextItemWidths we use such spacing to
contribute to the related InlineTextItem's width.

Important thing to notice:
Given the example:
<span>agua</span><span>水<span>
This would be represented by:
 InlineStartBox | InlineTextItem | InlineEndBox | InlineStartBox | InlineTextItem | InlineEndBox
 Being "agua" the content of the first InlineTextItem and "水" the content of the second InlineTextItem.

 Visually, the spacing necessary between "agua" and "水" has to appear between the first InlineEndBox
 and the second InlineStartBox. This means that such spacing should not visually contribute for
 the width of InlineTextItem. It might seem that we are doing this at this patch, however,
 this part is just addressing layout, and this information is sufficient for InlineContentBreaker to
 take such spacing in consideration. Later, on a future patch which will address "display", we will
 then need to place this spacing in its correct visual order: it should me moved in front of the InlineStartBox
 and its value should be removed from the InlineTextItem's width, such that it doesn't get accounted for twice.

[1] https://www.w3.org/TR/css-text-4/#text-autospace-property

* Source/WebCore/layout/formattingContexts/inline/InlineContentCache.h:
(WebCore::Layout::InlineContentCache::inlineBoxBoundaryTextSpacings const):
(WebCore::Layout::InlineContentCache::setInlineBoxBoundaryTextSpacings):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.h:
(WebCore::Layout::InlineFormattingContext::inlineContentCache const):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::build):
(WebCore::Layout::InlineItemsBuilder::adjustInlineBoxBoundaryTextSpacingsIfNeeded):
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.h:
* Source/WebCore/platform/text/TextSpacing.cpp:
(WebCore::TextAutospace::shouldApplySpacing const):
* Source/WebCore/platform/text/TextSpacing.h:

Canonical link: https://commits.webkit.org/282917@main
